src/main/java/org/spongepowered/api/item/inventory/menu/InventoryMenu.java

```diff
@@ -0,0 +1,167 @@
+     *
+     * @return the current inventory
+     */
+    ViewableInventory getCurrentInventory();
+
+    /**
+     * Returns the container type of the current inventory.
+     *
+     * @return current container type.
+     */
+    ContainerType getType();
+
+    /**
+     * Sets a new inventory. If the ContainerType does not change the inventory will be swapped out silently.
+     * <p>If the ContainerType is different all existing callbacks are cleared and open menus are closed and reopened with the new inventory.</p>
+     *
+     * @param inventory the new inventory
+     */
+    void setCurrentInventory(ViewableInventory inventory);
+
+    /**
+     * Sets the title of this menu.
+     * <p>Any open menus are closed and reopened with the new title.</p>
+     *
+     * @param title the new title.
+     */
+    void setTitle(Text title);
+
+    /**
+     * Registers click callbacks. Unregisters previously registered handler of the same type.
+     * <p>Return false in the handler to prevent changes.</p>
+     * <p>Possible handlers: {@link ClickHandler}, {@link SlotClickHandler}, {@link KeySwapHandler}, {@link SlotChangeHandler}, {@link CloseHandler}</p>
+     *
+     * @param handler the handler
+     */
+    void registerHandler(Object handler);
+
+    /**
+     * Registers a click callback. Unregisters previously registered handler of the same type.
+     * <p>Return false in the handler to prevent changes.</p>
+     *
+     * @param handler the handler
+     */
+
+    void registerClick(ClickHandler handler);
+
+    /**
+     * Registers a slot click callback. Unregisters previously registered handler of the same type.
+     * <p>Return false in the handler to prevent changes.</p>
+     *
+     * @param handler the handler
+     */
+    void registerSlotClick(SlotClickHandler handler);
+
+    /**
+     * Registers a key-swap click callback. Unregisters previously registered handler of the same type.
+     * <p>Return false in the handler to prevent changes.</p>
+     *
+     * @param handler the handler
+     */
+    void registerKeySwap(KeySwapHandler handler);
+
+    /**
+     * Registers a change callback. Unregisters previously registered SlotChangeHandler.
+     * <p>You can override the behaviour of {@link #setReadOnly(boolean)} with this.</p>
+     * <p>Return false in the handler to prevent changes.</p>
+     *
+     * @param handler the callback handler
+     */
+    void registerChange(SlotChangeHandler handler);
+
+    /**
+     * Registers a callback for when this menu is closed. Unregisters previously registered handler.
+     *
+     * @param handler the callback handler
+     */
+    void registerClose(CloseHandler handler);
+
+    /**
+     * Registers a callback handler that prevents any change
+     * <p>You can override the behaviour of {@link #setReadOnly(boolean)} with this.</p>
+     */
+    default void registerReadOnly() {
+        registerChange((container, slot, slotIndex) -> false);
```

Comment:
`setReadOnly(false)`?

Reply:
I had it with slotIndex filter before (also allowing multiple handlers). 
But it became too complex in impl.
So this will be gone completly.
`setReadonly(false)` does the same